crates/ruff_db/src/diagnostic/render.rs

```diff
@@ -67,62 +71,161 @@ impl std::fmt::Display for DisplayDiagnostic<'_> {
-
+                if let Some(span) = self.diag.primary_span() {
                     write!(
                         f,
-                        ":{line}:{col}",
-                        line = fmt_styled(start.line, stylesheet.emphasis),
-                        col = fmt_styled(start.column, stylesheet.emphasis),
+                        " {path}",
+                        path = fmt_styled(span.file().path(self.resolver), stylesheet.emphasis)
                     )?;
+                    if let Some(range) = span.range() {
+                        let diagnostic_source = span.file().diagnostic_source(self.resolver);
+                        let start = diagnostic_source
+                            .as_source_code()
+                            .line_column(range.start());
+
+                        write!(
+                            f,
+                            ":{line}:{col}",
+                            line = fmt_styled(start.line, stylesheet.emphasis),
+                            col = fmt_styled(start.column, stylesheet.emphasis),
+                        )?;
+                    }
+                    write!(f, ":")?;
+                }
+                writeln!(f, " {message}", message = self.diag.concise_message())?;
+            }
+            DiagnosticFormat::Azure => {
+                let severity = match self.diag.severity() {
+                    Severity::Info | Severity::Warning => "warning",
+                    Severity::Error | Severity::Fatal => "error",
+                };
+                write!(f, "##vso[task.logissue type={severity};")?;
+                if let Some(span) = self.diag.primary_span() {
+                    let filename = span.file().path(self.resolver);
+                    write!(f, "sourcepath={filename};")?;
+                    if let Some(range) = span.range() {
+                        let location = if self.resolver.notebook_index(span.file()).is_some() {
+                            // We can't give a reasonable location for the structured formats,
+                            // so we show one that's clearly a fallback
+                            LineColumn::default()
+                        } else {
+                            span.file()
+                                .diagnostic_source(self.resolver)
+                                .as_source_code()
+                                .line_column(range.start())
+                        };
+                        write!(
+                            f,
+                            "linenumber={line};columnnumber={col};",
+                            line = location.line,
+                            col = location.column,
+                        )?;
+                    }
+                }
+                writeln!(
+                    f,
+                    "{code}]{body}",
+                    code = self
+                        .diag
+                        .secondary_code()
```

Comment:
Same as for JSON. We can't use the `secondary_code` for ty. Instead, it should be the `id`.